firmware/shared/src/hw/hw_sdio_sd.c

```diff
@@ -0,0 +1,57 @@
+// interface read  write
+// using HAL layer
+#include "hw_sdio_sd.h"
+#include <stdlib.h>
+
+SD_card *hw_sdio_sd_init(SD_card_init_config config)
+{
+    SD_card *sd =
+        malloc(sizeof(SD_card)); // maybe replaced by other memory allocation function or don't do malloc at all
+
+    if (sd == NULL)
+    {
+        return NULL;
+    }
+
+    sd->hsd->Instance = config.sdio;
+    sd->hsd->Init     = config.sdio_config;
+    sd->timeout       = config.timeout;
+
+    // will init and configure the SDIO peripheral
+    // will use SDIO to send command to figure out the card information
+    HAL_StatusTypeDef status = HAL_SD_Init(sd->hsd);
+    if (status != HAL_OK)
+    {
+        return NULL;
+    }
+
+    return sd;
```

Comment:
I believe this is all handled by the init function in `main.c`. We might not need an init function at all, but you should check me on this.